### PR TITLE
Bug 1995727: fix tests, code around processing build,buildconfig deletes while PV based Jenkins is down

### PR DESCRIPTION
**Pull request by gabemontero**
See https://github.com/openshift/release/pull/21048#issuecomment-902019733, https://github.com/openshift/release/pull/21048#issuecomment-902039578, and https://github.com/openshift/release/pull/21048#issuecomment-902105002

summary:  existing e2e-aws-jenkins had a hole in its test for covering deleted builds in between jenkins restart where jenkins uses a PV for its state

/assign @coreydaley 
/assign @akram

### CODE REVIEW

**Comment by gabemontero** [comment]
/hold

running the PV scenario manually .... it is possible the deletes the test are doing are getting caught before the jenkins instance is shut down

**Comment by gabemontero** [comment]
note: I intentionally did not update the global build/builconfig watches to do the list, as there are no jenkins specific labels, only annotations, so we can't use label selectors to filter

but of course let's discuss during the review process

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java:347]: can we use a constant for this?

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java:155]: is it  enough to just log on Exception?
If yes, can you add a comment in code to explain why?

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java:145]: the `deleteInProgress` methods takes as a signature the `BuildConfig` name. Is there any reason to prepend namespace ?

**Review by akram — CHANGES_REQUESTED** [review]

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildManager.java:347]: turns our we already had one ;-) .... switching to that in next push

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java:155]: will add comment .... but yes, I believe logging and continuing to next BC vs. throwing exception and leaving method is best

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java:145]: I patterned off of https://github.com/openshift/jenkins-sync-plugin/blob/c6c5958a3f028112f24545a196f114342b2116a3/src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java#L94-L95 where both namespace and name are joined for the parameter supplied

also the corresponding deleteCompleted at https://github.com/openshift/jenkins-sync-plugin/blob/c6c5958a3f028112f24545a196f114342b2116a3/src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java#L100-L101 does that

finally, in general, we just need to account for the ns and name as 2 bc's in different namespaces could have the same name

Pretty sure this is right.

**Comment by gabemontero** [comment]
ok @akram pushed updates made based on your changes, posted explanation in comments for where I answered your question but left code unchanged

also pushed a separate commit to employ same changes on cluster level build/buildconfig watcher per our discussion

PTAL when you are next at work (most likely Monday)

thanks

**Comment by gabemontero** [comment]
my commits today based on responding to feedback/conversations with @akram appear to have introduced a bug

I'm seeing analogous failures here with e2e-aws-jenkins as well as with the analogous test in the new in repo e2e's running locally against my cluster where the jenkins imagestream is patched with this PR's version of the plugin

don't know why yet, but my initial educated guess is the update for the cluster wide build watcher introduced an error somehow

will start investigating, see if I can sort it out before @akram looks at these changes on Monday.

**Comment by gabemontero** [comment]
had a typo on the switch to the constant for the label selector ... sorted that out, tests passing locally again

also realized @akram that I didn't need to do any list calls on the informer starts nor call `onInit` any more ... instead I'm employing static initializer based calls to the reconciler methods that work independent of cluster scope or namespace scoped informers .... much cleaner and more efficient.

PTAL on Monday and we'll move forward on this.  Thanks.

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java
Comment (akram) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigManager.java:149]: ok, got it, I didn't check the other calls.
as an improvement, we should refactor this method then to set 2 parameters: `String namespace, String buildName`  or even better: juste the `BuildConfig bc` (as bc.metadata has reference to ns) and use a separator in the key map. 
Otherwise, you could face a crazy scenario if a user have this:
```
namespace1: java
bc1: build

namespace2: jav
bc2: abuild
```

Very poor likely to happen though.
I can push that RFE right after we merge this PR.

**Comment by akram** [comment]
> had a typo on the switch to the constant for the label selector ... sorted that out, tests passing locally again
> 
> also realized @akram that I didn't need to do any list calls on the informer starts nor call `onInit` any more ... instead I'm employing static initializer based calls to the reconciler methods that work independent of cluster scope or namespace scoped informers .... much cleaner and more efficient.
> 
> PTAL on Monday and we'll move forward on this. Thanks.

great to read this. 
static initializer should be ok also. 
I have reviewed the other changes and replies and that looks good to me.

**Pull request merged**
